documentation/design/proposals/global-console-phase-2.adoc

```diff
@@ -86,22 +86,25 @@ To signal a query across all namespaces, a * is used in the path name element.
 
 ==== exposing metrics
 
-The console-server will accept configuration that defines the statistics avaiable for each configured resource type.  This will take the form a list of named PromQL queries that are associated with each resource type (e.g. address).  These PromQL queries will ultimately produce the values that we wish to display on the console.  The idea is to avoid the need for the console UI to compute values itself.  
+The console-server will expose a endpoint which will proxy the to Prometheus PromQL api/v1/query API https://prometheus.io/docs/prometheus/latest/querying/api/.
```

Comment:
Thinking from a monitoring perspective, can this endpoint easily be scraped by prometheus or does some custom code have to be written in order to fetch the metrics?

Reply:
Originally I had thought that the backend would associate a list of named queries with each resource type.  The UI could then request these named queries and use the results to populate the display.  The difficulty I faced was that I needed these queries to be a parameterised.   There seemed little point in introducing another API layer wrapping PromQL, so it seemed simpler to me to expose Prometheus itself as an authenticated proxy.

Orthogonal point - I was thinking that we'd just employ the prometheus operator to deploy the prometheus instance used by the console.  Is there a reason to disallow the user from interacting with this prometheus instance (and maybe grafana too) directly?

Reply:
> Orthogonal point - I was thinking that we'd just employ the prometheus operator to deploy the prometheus instance used by the console. Is there a reason to disallow the user from interacting with this prometheus instance (and maybe grafana too) directly?

As long as its via a proxy that restricts which address spaces a user can get metrics from, I don't see a big problem with that. However, it does feel more natural to not expose prometheus directly as it is an implementation-specific part of the architecture.

I don't think there is a need to deploy Grafana?

Reply:
If we need to restrict access to metrics by address space, then I think the idea to expose PromQL doesn't work well.  The console-server would need to intercept the queries and ensure that no metric would be considered or returned that falls outside the address space(s) the user can see.   That implementation sounds unappealing.

Reply:
@lulf  I do have alternative idea.   At the moment, the console UI consumes the kubernetes addressspace API directly and the intention was to extend that for addresses and connections (latter kubernetes like).   As an alternative, what if we had the console-api exposed a different API for list/get operations.  The response would be an envelope combining  kubernetes object and the metrics.

 The console UI would still mutate using the existing kubernetes API but all reads would be done through this alternative.

The list of metrics included in each response envelope would be held in static configuration, with a list a separate list of PromQL queries held for each object type.

This means that the console-ui won't need to perform a separate operation to get the metrics, nor will it need to join the two results set together.

A addressspace list response would look like this:

```
{
  "items": [
    {
      "addresSpace": {
        "apiVersion": "enmasse.io/v1beta1",
        "kind": "AddressSpace",
        "metadata": {},
        "spec": {}
      },
      "metrics": [
        {
          "metricName": "enmasse_addresses",
          "prometheusMetricType": "guage",
          "prometheusMetricValue": 10
        },
        {
          "metricName": "enmasse_connections",
          "prometheusMetricType": "guage",
          "prometheusMetricValue": 10
        },
        {
          "metricName": "enmasse_messages_in_total",
          "prometheusMetricType": "counter",
          "prometheusMetricValue": 20
        }
      ]
    },
    {
      "addresSpace": {
        "apiVersion": "enmasse.io/v1beta1",
        "kind": "AddressSpace",
        "metadata": {},
        "spec": {}
      },
      "metrics": [
        {
          "metricName": "enmasse_addresses",
          "prometheusMetricType": "guage",
          "prometheusMetricValue": 12
        }
      ]
    }
  ]
}
```

in addition, there would be an API that would provide the metrics details for addresses and connections (i.e. the metrics for the senders/receivers).

Reply:
Yeah I think that could work. I don't think there is a big reason why we need to expose the same API as Kubernetes. We should expose whatever makes sense for this use case. However, to not cause too much changes, maybe separating the read and write path is ok.

FWIW this is where I believe GraphQL would fit very well, giving the client the ability to select which parts of the response that should be returned. I.e. not returning metrics, or just a specific metric.

Reply:
My original inclination was to avoid wrapping/transforming the kubernetes objects.  I was driven by the Console's future need to support other CRDs types (Strimizi etc). Keeping the UI speaking Kubernetes API seemed like a good approach. However, that approach leads you to a separate API for querying metrics and the need for the UI to join the two results set.